### PR TITLE
1186 evaluate loader for runner

### PR DESCRIPTION
**Pull request by leoromanovich**
## Before submitting (checklist)

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [x] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

## Description
Add evaluation_loader method to runner.

## Related Issue
https://github.com/catalyst-team/catalyst/issues/1186

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,19 @@ runner.train(
...
+metrics = runner.evaluate_loader(
+    model=runner.model, # or any other model, compatable with current runner
+    loader=loaders['valid'],
+    callbacks=[
+    dl.AccuracyCallback(
+        input_key="logits", target_key="targets", topk_args=(1, 3, 5)),
+        # ... any other callback, compatable with runner
+    ]
+    )
+assert 'accuracy' in metrics.keys()
```
Comment (Scitator) [inline README.md]: could you please run `catalyst-make-codestyle`on this block?

File: catalyst/runners/runner.py [hunk 2]
```diff
@@ -670,6 +670,52 @@ def on_loader_end(self, runner):
         for batch in loader:
             yield self.predict_batch(batch)
 
+    def evaluate_loader(
+        self,
+        model: Model,
```
Comment (Scitator) [inline catalyst/runners/runner.py]:
```suggestion
        model: Model = None,
```
we should not respecify the model after `.train`, isn't it?

File: README.md (hunk 1 above)
Comment (Scitator) [reply README.md]: just want the code to be the same style

**Review by Scitator — COMMENTED** [review]
overall, looks good to me

File: README.md (hunk 1 above)
Comment (leoromanovich) [reply README.md]: Sure.) 
I've added a changed by catalyst-codestyle block to index.rst too.

File: catalyst/runners/runner.py (hunk 2 above)
Comment (Scitator) [reply catalyst/runners/runner.py]: @leoromanovich could you check this out too?

Comment (leoromanovich) [reply catalyst/runners/runner.py]: Yeah. 
Do you mean, that if a model is None (only callbacks and loader are given), we need simply take the model from the runner state? 
Like:
```python
if model is None:
    model = self.model
```

Comment (leoromanovich) [reply catalyst/runners/runner.py]: @Scitator In the current version, if a user explicitly changes the model, it will affects on a global state (for example, if after that user runs `predict_loader`, a `runner` will use the model given in `evaluate_loader`). I'm not sure if it's proper behavior. 
What do you think?

Comment (Scitator) [reply catalyst/runners/runner.py]: For now, every time you pass a new model to the Runner parameters, it wraps its behavior, so it looks okay.
Nevertheless, that's why I suggest placing `model=None` by default and check `self.model is not None` to drive a user to a one-time model specification for the Runner (we have an init option, btw :) )

Comment (leoromanovich) [reply catalyst/runners/runner.py]: I've added this to runner and fixed docs wrt it =)

**Comment by penguinflys** [comment]
May I ask, why is ``CheckpointCallback isn`t allowed for evaluation loader method`` ?

File: catalyst/runners/runner.py [hunk 3]
```diff
@@ -670,6 +670,56 @@ def on_loader_end(self, runner):
...
+                if isinstance(callback, CheckpointCallback):
+                    raise RunnerException(
+                        "CheckpointCallback isn`t allowed for evaluation loader method"
+                    )
+        else:
+            for callback in callbacks.values():
+                if isinstance(callback, CheckpointCallback):
+                    raise RunnerException(
+                        "CheckpointCallback isn`t allowed for evaluation loader method"
+                    )
```
Comment (jyuan-dev) [inline catalyst/runners/runner.py:706]: May I ask, Why is CheckpointCallback not allowed for evaluation?

Comment (leoromanovich) [reply catalyst/runners/runner.py:706]: @penguinflys, thanks for the question! 
I think the idea behind evaluation_loader functionality is to minimize behavior to "make a loader inference and return metrics, not more." In the beginning, we decided to reduce functionality because of the "good first issue" label =) 
If you have any suggestions, feel free to share them!

Comment (jyuan-dev) [reply catalyst/runners/runner.py:706]: I think `CheckpointCallback` could be very helpful if the user makes a prediction loading the trained model.

Comment (Scitator) [reply catalyst/runners/runner.py:706]: `resume` feature is under development
we have a few questions to solve before its an introduction
it's a bit complicated due to the fact, that user could use any possible PyTorch checkpoints ;)
for such a reason loading the model is out of scope for now

Comment (jyuan-dev) [reply catalyst/runners/runner.py:706]: thanks for feedbacking
